### PR TITLE
Add support to the "map" validator for constraining the keys

### PR DESCRIPTION
**Pull request by Phidica**
Per discussion in #53 and #62.

Accepting this pull request would add support for a `key =` constraint on the `map` validator that allows another validator to be applied to the keys of the mapping.

I have written some tests, and I am open to writing more test cases if needed. I almost feel like this was *too* simple to implement, so I fear there must be some nasty corner case problems that are yet unsolved.

I will gladly add more commits to update documentation later. 
For now, I am looking for discussion and input on the implementation :)

### CODE REVIEW

File: yamale/validators/base.py [hunk 1]
```diff
@@ -62,7 +62,10 @@ def validate(self, value):
         for constraint in self._constraints_inst:
             error = constraint.is_valid(value)
             if error:
-                errors.append(error)
+                if isinstance(error, list):
+                    errors.extend(error)
```
Comment (Phidica) [inline yamale/validators/base.py:66]: Of course, the `error` object returned by the `Key` constraint test could be a string just like all the other constraints do, rather than a list as I have done. However, I discovered that the tests in the `Key` constraint could be performed in a different order on Python 2.7 vs Python 3.7 (my version) because dictionaries are now guaranteed to remain in insert order, while before they were not. This meant the constructed string could be totally different between different Python versions (because errors may have been encountered in a different order), and this made it difficult to assert the value of this error output in a version-agnostic way. So, I felt a better solution was to extend the total list of errors with each individual error from the `Key` constraint, because this final list is tested in an arbitrary order by the testing suite. It is certainly a sort of workaround to a more fundamental problem though.

**Comment by mildebrandt** [comment]
Awesome, thanks! It may take a little bit for me to get to this, please be patient.

File: yamale/validators/base.py (hunk 1 above)
Comment (mildebrandt) [reply yamale/validators/base.py:66]: So, @abourree has a solution to that here: https://github.com/23andMe/Yamale/pull/83/files#diff-cbb0754bd27ec59a69abd0f651e48c3bR290

The `match_exception_lines` method was changed to sort the two lists of errors before comparison. I'd rather keep that. I think the best thing to do is to copy over the changes to that method into this PR since #83 is a bigger piece of work and will take longer to get in. 

Sorry, one of you will be getting a merge conflict to resolve depending whose gets in first.

Comment (mildebrandt) [reply yamale/validators/base.py:66]: Oh, I may have misunderstood the issue. The change in #83 is about checking all the errors in multiple validators/constraints and you're concerned with returning multiple errors in your one constraint. This seems fine then, sorry for the confusion.

**Comment by mildebrandt** [comment]
This looks really good. Can you please update the README? Thanks!

**Comment by Phidica** [comment]
@mildebrandt How's this?

**Comment by mildebrandt** [comment]
Thanks for your contribution!
